### ISSUE TITLE
Cannot use DDP to train the segmentation model

### ISSUE THREAD

**Issue by tianlan6767** [ultralytics/ultralytics#1457]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

BUG
 when i train a v8 segment model, I got error as follows，Can you please help me take a look at this issue? Thank you very much.
![image](https://user-images.githubusercontent.com/71381036/225584113-b3f27c58-4dad-4d42-ab0c-56a33563c14e.png)

   
Environment:
    1.Ultralytics YOLOv8.0.51
    2.python3.7
    3.torch1.9.1+cu111
    4.RTX3090

PYTHON  TRAIN COMMAND  :
  -  python ultralytics/ultralytics/yolo/v8/segment/train.py

USE YAML：
     
![image](https://user-images.githubusercontent.com/71381036/225579417-e22c36a4-e8b2-4755-8106-a6c6fce4d626.png)

BUG:
   
![image](https://user-images.githubusercontent.com/71381036/225579949-88cca7a1-e144-42ca-be41-f8fa19b28061.png)
![image](https://user-images.githubusercontent.com/71381036/225580170-01aea646-598b-4d00-9c30-3aff26aeb871.png)

**Comment by AyushExel** [comment]
Please use the yolo command for training:
```
yolo segment train device="1,2,3,4"
```
https://docs.ultralytics.com/tasks/segment/

**Comment by tianlan6767** [comment]
> Please use the yolo command for training:
> 
> ```
> yolo segment train device="1,2,3,4"
> ```
> 
> https://docs.ultralytics.com/tasks/segment/


@AyushExel, thank u for your reply. i use the yolo command( yolo segment train  data=/ultralytics/datasets/OQC.yaml model=/ultralytics/models/v8/seg/yolov8m-seg.yaml pretrained=True epochs=300 imgsz= 2048 batch = 2 device=0,1) ,i still got the error
![image](https://user-images.githubusercontent.com/71381036/225797574-f7b1eb1f-f72e-419e-b587-db7a40280f05.png)

**Comment by Mason1992-Git** [comment]
the same question in 3090.i use python train wtih model.train('./yolov8s.pt',device='0,1')

**Comment by Laughing-q** [comment]
@Mason1992-Git @tianlan6767 hi guys, you might encounter the case that there're no objects in one batch while training, which would lead a `unused gradient` error when you're training with DDP mode because the proto module in segment head is not returning gradients.
Please uncomment the following comments if you encounter that error:
[https://github.com/ultralytics/ultralytics/blob/840bc689f7f4497dcc670078af65adf9165ef4f7/ultralytics/yolo/v8/segment/train.py#L123-L127](https://github.com/ultralytics/ultralytics/blob/840bc689f7f4497dcc670078af65adf9165ef4f7/ultralytics/yolo/v8/segment/train.py#L123-L127)
Actually the `no objects in one batch` case rarely happens and the solution code looks a little odd so we just leave the code commented there. :)

**Comment by Mason1992-Git** [comment]
您好，您的来信已收到，我会尽快阅读并回信。谢谢。袁磊

**Comment by glenn-jocher** [comment]
@Laughing-q do you think we should uncomment this section?

**Comment by glenn-jocher** [comment]
@Mason1992-Git @tianlan6767 I've implemented a fix for this bug in https://github.com/ultralytics/ultralytics/pull/1475/commits/ee49d1efd4bc6bcfedc59b0c31217a63b7c7e2fc. Can you please test out this branch and see if this resolves the problem for you? You can use this code to check out the 'updates' branch that contains the fix commit:

```bash
git clone https://github.com/ultralytics/ultralytivs -b updates
pip install -e ultralytics

yolo train ...
```

**Comment by Mason1992-Git** [comment]
> @Mason1992-Git @tianlan6767 I've implemented a fix for this bug in [ee49d1e](https://github.com/ultralytics/ultralytics/commit/ee49d1efd4bc6bcfedc59b0c31217a63b7c7e2fc). Can you please test out this branch and see if this resolves the problem for you? You can use this code to check out the 'updates' branch that contains the fix commit:
> 
> ```shell
> git clone https://github.com/ultralytics/ultralytics -b updates
> pip install -e ultralytics
> 
> yolo train ...
> ```

ok,i will try with it and reply later

**Comment by tianlan6767** [comment]
> @Mason1992-Git @tianlan6767 I've implemented a fix for this bug in [ee49d1e](https://github.com/ultralytics/ultralytics/commit/ee49d1efd4bc6bcfedc59b0c31217a63b7c7e2fc). Can you please test out this branch and see if this resolves the problem for you? You can use this code to check out the 'updates' branch that contains the fix commit:
> 
> ```shell
> git clone https://github.com/ultralytics/ultralytics -b updates
> pip install -e ultralytics
> 
> yolo train ...
> ```

thanks， i will try it as soon as possible

**Comment by tianlan6767** [comment]
> 

@Laughing-q but i don't understand, every image in the training set has objects, can you tell me why does it happen that no objects in one batch

**Comment by glenn-jocher** [comment]
@tianlan6767 great, thanks for letting us know the fix works!
